### PR TITLE
Record relative dynamic-require paths

### LINKED ISSUE #7339: Dynamic require()'s generate records that vary on different machines
# Bug report

**What is the current behavior?**

If you enable recordsPath and use a dynamic require (eg `require("./dependencies/" + dependency)`), the resulting line in your recordsPath file looks like:

```
    "src/dependencies sync ../../../../^/.//.*$": "./src/dependencies sync recursive ^\\.\\/.*$"
```

AFAICT, the number of `../`s corresponds to the depth of your project in your local filesystem (on mine, that file was built from `/Users/jon/Developer/require-test`).

This seems undesirable - it means that if you're version controlling recordsPath and build it on a different machine, a new record will be generated depending on the depth of the project on that other filesystem.

**If the current behavior is a bug, please provide the steps to reproduce.**

Checkout https://github.com/jdelStrother/webpack-require-test, built it.  Move it to a different directory (at a different depth in the filesystem).   Build it, note how webpack-records.json has changed.

**What is the expected behavior?**

The records should be the same regardless of where the project was built from

**Other relevant information:**
webpack version:  4.8.3
Node.js version: 8.4.0
Operating System: OS X 10.13.5
Additional tools:

### PR DESCRIPTION
**Pull request by jdelStrother**
**What kind of change does this PR introduce?**

This avoids records like:

`"src/dependencies sync ../../../../^/.//.*$": "./src/dependencies sync recursive ^\\.\\/.*$"`

when using dynamic requires (eg `require("./foo/" + file)`), with the number of "../"s varying according to the depth of the folder you're building from.

Fixes #7339

**Did you add tests for your changes?**

Yes

**Does this PR introduce a breaking change?**

Don't believe so.

**What needs to be documented once your changes are merged?**

Nothing(?)

---

WDYT to this?

### CODE REVIEW

File: lib/util/identifier.js [hunk 1]
```diff
@@ -12,6 +12,11 @@ const path = require("path");
  * @returns {boolean} returns true if path is "Absolute Path"-like
  */
 const looksLikeAbsolutePath = maybeAbsolutePath => {
+	if (maybeAbsolutePath === "/^\\.\\/.*$/") {
```
[the remark this answers is not in the document]
Comment (jdelStrother) [reply lib/util/identifier.js]: Actually, no.   We could also get stringified regexps here from something like:
```js
require.context("./dependencies", false, /\.js$/);
```

Maybe I should just check if it starts and ends with a `/`, to determine if its regexp-y.  I couldn't find any cases where an absolute filepath would be passed in here with a trailing slash (eg `"/src/"`) - what do you think?

File: test/configCases/records/issue-7339/test.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+function loadDependency(dep) {
+  require("./dependencies/" + dep)
+}
+
+it("should write relative dynamic-require paths to records", function() {
+	var fs = require("fs");
+	var path = require("path");
+	var content = fs.readFileSync(path.join(__dirname, "records.json"), "utf-8");
+	expect(content).toEqual(`{
```
[the remark this answers is not in the document]
Comment (jdelStrother) [reply test/configCases/records/issue-7339/test.js]: I could do... you're sure about that, though?  None of the other configTestCases use snapshots AFAICT - I was trying to mimic, eg, https://github.com/webpack/webpack//blob/master/test/configCases/records/issue-2991/test.js#L2.

File: lib/util/identifier.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sokra) [reply lib/util/identifier.js]: Threading strings starting and ending with `/` as RegExp, seems to be safe in my opinion. Every other absolute path is a resolved path and doesn't end with `/`.

File: test/configCases/records/issue-7339/test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sokra) [reply test/configCases/records/issue-7339/test.js]: I would also say a snapshot would be appropriate here. The other test was written before we used jest and we did have access to snapshot tests these days. If you want, you can change the other test to a snapshot too.

**Comment by sokra** [comment]
Thanks

**Pull request merged**